src/kbc_modules/eaa_kbc/mod.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::kbc_modules::{KbcCheckInfo, KbcInterface};
+use anyhow::*;
+use log::*;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::net::TcpStream;
+use std::os::unix::io::AsRawFd;
+
+pub mod protocol;
+pub mod rats_tls;
+use protocol::*;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct AnnotationPacket {
+    pub kid: String,
+    pub wrapped_data: Vec<u8>,
+    pub iv: Vec<u8>,
+    pub algorithm: String,
+    pub key_length: u16,
+}
+
+pub struct EAAKbc {
+    pub kbs_uri: String,
+    pub protocol_version: String,
+    pub algorithm: String,
+    pub key_length: u16,
+    // pub kek_cache: HashMap<String, String>,
+    pub tcp_stream: Option<TcpStream>,
+    pub tls_handle: Option<rats_tls::RatsTls>,
+}
+
+impl KbcInterface for EAAKbc {
+    fn check(&self) -> Result<KbcCheckInfo> {
+        let mut kbs_info: HashMap<String, String> = HashMap::new();
+        kbs_info.insert("kbs_addr".to_string(), self.kbs_uri.clone());
+        kbs_info.insert(
+            "protocol_version".to_string(),
+            self.protocol_version.clone(),
+        );
+        Ok(KbcCheckInfo { kbs_info: kbs_info })
+    }
+
+    fn decrypt_payload(&mut self, annotation: &str) -> Result<Vec<u8>> {
+        debug!("EAA KBC decrypt_payload() is called!");
+        let annotation_packet: AnnotationPacket = serde_json::from_str(annotation)?;
+        self.algorithm = annotation_packet.algorithm;
+        self.key_length = annotation_packet.key_length;
+        if self.tcp_stream.is_none() {
+            debug!("First request, connecting KBS...");
+            self.establish_new_kbs_connection()?;
+            debug!("connect success! TLS is established!");
+        }
+
+        debug!("start decrypt...");
+        let decrypted_payload = self.kbs_decrypt_payload(
+            annotation_packet.wrapped_data,
+            annotation_packet.kid,
+            annotation_packet.iv,
+        )?;
+        debug!("decrypted success!");
+        Ok(decrypted_payload)
+    }
+}
+
+impl EAAKbc {
+    pub fn new(kbs_uri: String) -> EAAKbc {
+        EAAKbc {
+            kbs_uri: kbs_uri,
+            protocol_version: String::new(),
+            algorithm: String::new(),
+            key_length: 0,
+            // kek_cache: HashMap::new(),
+            tcp_stream: None,
+            tls_handle: None,
+        }
+    }
+
+    fn establish_new_kbs_connection(&mut self) -> Result<()> {
+        debug!("create RATS TLS handle...");
+        self.tls_handle = Some(
+            rats_tls::RatsTls::new(
+                false,
+                0,
```

Comment:
This is the enclave_id? Why are we always setting it to zero?

Reply:
Rats-tls first supports the remote authentication of SGX. This enclave ID is mainly used in SGX scenarios. This parameter is not used in TDX/SEV confidential computing scenarios, so it is set to zero here.

Reply:
ok cool. it might be handy to add a comment mentioning that server, enclave_id, and mutual are all unused.